flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java

```
@@ -0,0 +1,197 @@
+
+	private final SchedulingTopology schedulingTopology;
+
+	private final Map<ExecutionVertexID, DeploymentOption> deploymentOptions;
+
+	private final InputDependencyConstraintChecker inputConstraintChecker;
+
+	public LazyFromSourcesSchedulingStrategy(
+			SchedulerOperations schedulerOperations,
+			SchedulingTopology schedulingTopology) {
+		this.schedulerOperations = checkNotNull(schedulerOperations);
+		this.schedulingTopology = checkNotNull(schedulingTopology);
+		this.deploymentOptions = new HashMap<>();
+		this.inputConstraintChecker = new InputDependencyConstraintChecker();
+	}
+
+	@Override
+	public void startScheduling() {
+		List<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions = new ArrayList<>();
+		final DeploymentOption updateOption = new DeploymentOption(true);
+		final DeploymentOption nonUpdateOption = new DeploymentOption(false);
+
+		for (SchedulingExecutionVertex schedulingVertex : schedulingTopology.getVertices()) {
+			DeploymentOption option = nonUpdateOption;
+			for (SchedulingResultPartition srp : schedulingVertex.getProducedResultPartitions()) {
+				if (srp.getPartitionType().isBlocking()) {
+					inputConstraintChecker.addSchedulingResultPartition(srp);
+				} else {
+					option = updateOption;
+				}
+			}
+			deploymentOptions.put(schedulingVertex.getId(), option);
+
+			if (schedulingVertex.getConsumedResultPartitions().isEmpty()) {
+				// schedule vertices without consumed result partition
+				executionVertexDeploymentOptions.add(
+						new ExecutionVertexDeploymentOption(schedulingVertex.getId(), option));
+			}
+		}
+
+		schedulerOperations.allocateSlotsAndDeploy(executionVertexDeploymentOptions);
+	}
+
+	@Override
+	public void restartTasks(Set<ExecutionVertexID> verticesToRestart) {
+		// increase counter of the dataset first
+		for (ExecutionVertexID executionVertexId : verticesToRestart) {
+			final SchedulingExecutionVertex schedulingVertex = schedulingTopology.getVertex(executionVertexId)
```

Code is duplicated in line 101. For now I think it's ok to introduce a helper method inside the strategy. Eventually we have to think whether `Optional` return type makes sense, or if `schedulingTopology` should just throw an exception`.
```
	private SchedulingExecutionVertex getSchedulingVertex(final ExecutionVertexID executionVertexId) {
		return schedulingTopology.getVertex(executionVertexId)
			.orElseThrow(() -> new IllegalStateException("can not find scheduling vertex for " + executionVertexId));
	}
```

---

Yes, I also think that `Optional` may be a little code redundant since `orElseThrow` is always checked. May be we could optimize this in another JIRA ticket.

---

I think we can add `getVertexOrThrow(ExecutionVertexID)` to the topology. Can you create a ticket for that?

---

OK. [FLINK-12608](https://issues.apache.org/jira/browse/FLINK-12608)